common/stash/src/main/java/com/bazaarvoice/emodb/common/stash/StashReader.java

```diff
@@ -67,20 +72,83 @@ protected static AmazonS3 getS3Client(URI stashRoot, AWSCredentialsProvider cred
+                .orElseGet(() -> Reflection.newProxy(AmazonS3.class, new AbstractInvocationHandler() {
+                    private AmazonS3 _resolvedClient = null;
+
+                    @Override
+                    protected Object handleInvocation(Object proxy, Method method, Object[] args) throws Throwable {
+                        return method.invoke(resolvedClient(), args);
+                    }
+
+                    private AmazonS3 resolvedClient() {
+                        if (_resolvedClient == null) {
+                            String endPoint = determineEndpointForBucket(bucket, credentialsProvider, s3Config, stashRoot.getPath());
+                            _resolvedClient = createS3ClientForEndpoint(endPoint, credentialsProvider, s3Config);
+                        }
+                        return _resolvedClient;
+                    }
+                }));
+    }
+
+    private static AmazonS3 createS3ClientForRegion(String region, AWSCredentialsProvider credentialsProvider,
+                                                    @Nullable ClientConfiguration s3Config) {
+        return createS3ClientForEndpoint(String.format("s3.%s.amazonaws.com", region), credentialsProvider, s3Config);
+    }
+
+    private static AmazonS3 createS3ClientForEndpoint(String endPoint, AWSCredentialsProvider credentialsProvider,
+                                                      @Nullable ClientConfiguration s3Config) {
         AmazonS3 s3;
         if (s3Config == null) {
             s3 = new AmazonS3Client(credentialsProvider);
         } else {
             s3 = new AmazonS3Client(credentialsProvider, s3Config);
         }
-        // Determine the region for the bucket
-        Region region = StashUtil.getRegionForBucket(stashRoot.getHost());
-        s3.setRegion(region);
+        s3.setEndpoint(endPoint);
         return s3;
     }
 
+    private static String determineEndpointForBucket(String bucket, AWSCredentialsProvider credentialsProvider,
+                                                     @Nullable ClientConfiguration s3Config, String rootPath) {
+
+        // Guess us-east-1.  If wrong AWS will return a redirect with the correct endpoint
+        AmazonS3 s3 = createS3ClientForRegion(Regions.US_EAST_1.getName(), credentialsProvider, s3Config);
+        if (rootPath.startsWith("/")) {
+            rootPath = rootPath.substring(1);
+        }
+        if (!rootPath.endsWith("/")) {
+            rootPath = rootPath + "/";
+        }
+
+        try {
+            // Any request will work but presumably the client has list access for stash so perform a list.
+            s3.listObjects(new ListObjectsRequest()
+                    .withBucketName(bucket)
+                    .withPrefix(rootPath)
+                    .withDelimiter("/")
+                    .withMaxKeys(1));
+
+            // If this didn't error out then the presumed us-east-1 region was correct
+            return  "s3.us-east-1.amazonaws.com";
```

Comment:
just a thought - this return should never happen as the Exception is always expected as we only come here for determining the endpoint if it's not our regular region. So, just curious if we are saying Ok return, then we are contradicting our code :? may be go with some random 3rd region in our assumption or explicitly mention that the exception code has to run :? just need your opinion....

Reply:
What's happening here is we don't know what the region is.  Say someone made their own stash in bucket "my-stash-bucket".  The Stash code doesn't know what region that's in.  So what this does is it tries with us-east-1.  If that's right no exception is thrown, our guess of us-east-1 was correct.  If that's wrong then the exception is thrown, in which case we'll get the endpoint for the correct bucket from the exception.